File: Discord_win/Discord_win/Tools/ImageResolver.cs
using Discord.Resources.Static;
using Peanut.Client;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Media.Imaging;

namespace Discord.Tools {
    public static class ImageResolver {
        private static ImageDownloader ImageDownloader { get; } = new ImageDownloader(FileSystem.UserDirectory);
        private static HashSet<string> SetDownloadingImage { get; } = new HashSet<string>();
        private static Dictionary<string, List<TaskCompletionSource<bool>>> Queue { get; } = new Dictionary<string, List<TaskCompletionSource<bool>>>();
        public static async Task DownloadUserImageAsync(string imageName, Action<BitmapImage> actionResult) {
            string imagePath = FileSystem.MakeUserImageFilePath(imageName);
            // The file is being downloaded
            if (SetDownloadingImage.Contains(imagePath)) {
                // But the waiting queue hasn't been initialized yet
                if (!Queue.Keys.Contains(imagePath)) {
                    // Initialize it!
                    Queue.Add(imagePath, new List<TaskCompletionSource<bool>>());
                }
                // Enqueue and wait until the file is completely downloaded
                List<TaskCompletionSource<bool>> waiters = Queue[imagePath];
                TaskCompletionSource<bool> waiter = new TaskCompletionSource<bool>();
                waiters.Add(waiter);
                // Wait!
                await waiter.Task;
                // Make bitmap and throw it out!
                actionResult?.Invoke(GetBitmapFromLocalFile(imagePath));
            }
            // The file is not being downloaded
            else {
                // But it doesn't exist as well
                if (!File.Exists(imagePath)) {
                    // Download it!
                    SetDownloadingImage.Add(imagePath);
                    Downloader downloader = ImageDownloader.MakeDownloader(imageName);
                    downloader.OnDone += (o, e) => {
                        SetDownloadingImage.Remove(imagePath);
                        // Done downloading and there is a queue waiting for this file
                        if (Queue.Keys.Contains(imagePath)) {
                            // Notify the waiters that the file is completely downloaded
                            List<TaskCompletionSource<bool>> waiters = Queue[imagePath];
                            foreach (TaskCompletionSource<bool> waiter in waiters) {
                                waiter.SetResult(true);
                            }
                            Queue.Remove(imagePath);
                        }
                        actionResult?.Invoke(GetBitmapFromLocalFile(imagePath));
                    };
                    await downloader.StartDownloadingAsync();
                }
                // And it does exist
                else {
                    // Make bitmap and throw it out!
                    actionResult.Invoke(GetBitmapFromLocalFile(imagePath));
                }
            }
        }
        public static BitmapImage GetBitmapFromLocalFile(string filePath) {
            if (!File.Exists(filePath)) {
                return null;
            }
            byte[] imageBytes = File.ReadAllBytes(filePath);
            MemoryStream memoryStream = new MemoryStream(imageBytes);
            BitmapImage bitmapImage = new BitmapImage();
            bitmapImage.BeginInit();
            bitmapImage.StreamSource = memoryStream;
            try {
                bitmapImage.EndInit();
                return bitmapImage;
            }
            catch (Exception) {
                return null;
            }
        }
    }
}

File: Discord_android/Discord/app/src/main/java/com/peanut/discord/tools/ImageResolver.java
package com.peanut.discord.tools;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.peanut.androidlib.common.client.Downloader;
import com.peanut.androidlib.common.worker.SingleWorker;
import com.peanut.discord.resources.Data;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
public class ImageResolver {
    private static ImageDownloader imageDownloader = new ImageDownloader();
    private static List<String> listDownloadingFile = new LinkedList<>();
    private static HashMap<String, List<Object>> queue = new HashMap<>();
    private static HashMap<String, Bitmap> mapFilePathBitmap = new HashMap<>();
    private static SingleWorker singleWorker = new SingleWorker();
    public static void downloadUserImage(String imageName, OnFinishListener onFinishListener) {
        if(imageName == null){
            return;
        }
        singleWorker.execute(() -> {
            String imagePath = Data.makeUserImageFilePath(imageName);
            if(mapFilePathBitmap.containsKey(imagePath)){
                onFinishListener.onFinish(mapFilePathBitmap.get(imagePath));
                return;
            }
            if(listDownloadingFile.contains(imagePath)){
                if(!queue.containsKey(imagePath)){
                    queue.put(imagePath, new ArrayList<>());
                }
                List<Object> waiters = queue.get(imagePath);
                Object waiter = new Object();
                waiters.add(waiter);
                synchronized (waiter){
                    try {
                        waiter.wait();
                        if(onFinishListener != null){
                            onFinishListener.onFinish(mapFilePathBitmap.get(imagePath));
                        }
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
            else{
                File file = new File(imagePath);
                if(!file.exists()){
                    listDownloadingFile.add(imagePath);
                    Downloader downloader = imageDownloader.makeDownloader(imageName);
                    downloader.setOnDoneListener(() -> {
                        mapFilePathBitmap.put(imagePath, makeBitmapFromLocalFile(imagePath));
                        listDownloadingFile.remove(imagePath);
                        if(queue.containsKey(imagePath)){
                            List<Object> waiters = queue.get(imagePath);
                            for(Object waiter : waiters){
                                synchronized (waiter){
                                    waiter.notify();
                                }
                            }
                            queue.remove(imagePath);
                        }
                        if(onFinishListener != null){
                            onFinishListener.onFinish(mapFilePathBitmap.get(imagePath));
                        }
                    });
                    downloader.start();
                }
                else{
                    if(mapFilePathBitmap.containsKey(imagePath)){
                        mapFilePathBitmap.put(imagePath, makeBitmapFromLocalFile(imagePath));
                    }
                    if(onFinishListener != null){
                        onFinishListener.onFinish(mapFilePathBitmap.get(imagePath));
                    }
                }
            }
        });

    }
    private static Bitmap makeBitmapFromLocalFile(String filePath){
        File file = new File(filePath);
        if(!file.exists() || !file.isFile()){
            return null;
        }
        return BitmapFactory.decodeFile(filePath);
    }
    public interface OnFinishListener {
        void onFinish(Bitmap bitmap);
    }
}